File: tests/test_timer.py
# This file is part of GreenMoon2D, a 2D game engine for Python
# written by Willi Kappler, MIT license.
#
# See: https://github.com/willi-kappler/green_moon_2d

import unittest
import time

from green_moon_2d.gm_timer import GMTimer


class TestVec2D(unittest.TestCase):
    def test_finished(self):
        """
        Test if timer calculates finished value correctly.
        """

        t1 = GMTimer(100)
        self.assertEqual(t1.finished(), False)

        time.sleep(0.055)
        self.assertEqual(t1.finished(), False)

        time.sleep(0.055)
        self.assertEqual(t1.finished(), True)

        t1.active = False
        self.assertEqual(t1.finished(), False)

        time.sleep(0.11)
        self.assertEqual(t1.finished(), False)

    def test_set_duration(self):
        """
        Test changing the duration of the timer.
        """

        t1 = GMTimer(200)

        t1.set_duration(500)
        self.assertEqual(t1.duration, 500)

    def test_restart(self):
        """
        Test restating the time once it has finished.
        """

        t1 = GMTimer(100)

        time.sleep(0.11)
        self.assertEqual(t1.finished(), True)

        t1.restart()
        self.assertEqual(t1.finished(), False)

        time.sleep(0.11)
        self.assertEqual(t1.finished(), True)

        t1.active = False
        self.assertEqual(t1.finished(), False)

        t1.restart()
        time.sleep(0.11)
        self.assertEqual(t1.finished(), True)

    def test_set_duration_restart(self):
        """
        Test changing the duration and restarting the timer.
        """

        t1 = GMTimer(100)
        t1.active = False

        self.assertEqual(t1.finished(), False)

        time.sleep(0.11)
        self.assertEqual(t1.finished(), False)

        t1.set_duration_restart(200)
        self.assertEqual(t1.active, True)
        self.assertEqual(t1.duration, 200)
        self.assertEqual(t1.finished(), False)

        time.sleep(0.21)
        self.assertEqual(t1.finished(), True)


if __name__ == "__main__":
    unittest.main()

File: tests/test_timer.cpp
/*
    GreenMoon2D
    SPDX-License-Identifier: MIT
    Written by Willi Kappler, MIT License
    https://github.com/willi-kappler/green_moon_2d

    This file contains the tests for the timer class.

    Run only configuration tests:
    xmake run -w ./ gm2d_test [timer]
*/

// STD includes:
#include <thread>

// External includes:
#include <snitch/snitch.hpp>

// Local includes:
#include "gm2d/gm_timer.hpp"

using namespace gm2d;

TEST_CASE("Create Timer", "[timer]" ) {
    GMTimer t1{};
    REQUIRE(t1.gm_finished());

    t1 = GMTimer(1000); // 1000 ms = 1s
    REQUIRE(!t1.gm_finished());

    auto const sleep_time = std::chrono::milliseconds(1000);
    std::this_thread::sleep_for(sleep_time);
    REQUIRE(t1.gm_finished());
}

TEST_CASE("Set timer duration", "[timer]" ) {
    GMTimer t1{};
    REQUIRE(t1.gm_finished());

    t1.gm_set_duration(1000);
    REQUIRE(!t1.gm_finished());

    auto const sleep_time = std::chrono::milliseconds(1000);
    std::this_thread::sleep_for(sleep_time);
    REQUIRE(t1.gm_finished());
}

TEST_CASE("Restart timer", "[timer]" ) {
    GMTimer t1{1000};
    REQUIRE(!t1.gm_finished());

    auto sleep_time = std::chrono::milliseconds(1000);
    std::this_thread::sleep_for(sleep_time);
    REQUIRE(t1.gm_finished());

    t1.gm_restart();
    REQUIRE(!t1.gm_finished());

    sleep_time = std::chrono::milliseconds(1000);
    std::this_thread::sleep_for(sleep_time);
    REQUIRE(t1.gm_finished());
}

TEST_CASE("Set timer duration restart", "[timer]" ) {
    GMTimer t1{1000};
    REQUIRE(!t1.gm_finished());

    auto sleep_time = std::chrono::milliseconds(1000);
    std::this_thread::sleep_for(sleep_time);
    REQUIRE(t1.gm_finished());

    t1.gm_set_duration_restart(500);
    REQUIRE(!t1.gm_finished());

    sleep_time = std::chrono::milliseconds(500);
    std::this_thread::sleep_for(sleep_time);
    REQUIRE(t1.gm_finished());
}

TEST_CASE("Set active", "[timer]" ) {
    GMTimer t1{1000};
    REQUIRE(!t1.gm_finished());

    auto sleep_time = std::chrono::milliseconds(1000);
    std::this_thread::sleep_for(sleep_time);
    REQUIRE(t1.gm_finished());

    t1.gm_set_active(false);
    REQUIRE(!t1.gm_finished());
    t1.gm_set_active(true);
    REQUIRE(t1.gm_finished());
}
